includes/functions.php

```diff
@@ -1483,24 +1486,35 @@ function fix_integer_value($value)
  * Find a device that has this IP. Checks ipv4_addresses and ipv6_addresses tables.
  *
  * @param string $ip
+ * @param array $within_poller_groups
  * @return \App\Models\Device|false
  */
-function device_has_ip($ip)
+function device_has_ip($ip, $within_poller_groups = [])
 {
     if (IPv6::isValid($ip)) {
-        $ip_address = \App\Models\Ipv6Address::query()
+        $ip_addresses = \App\Models\Ipv6Address::query()
             ->where('ipv6_address', IPv6::parse($ip, true)->uncompressed())
             ->with('port.device')
-            ->first();
+            ->get();
     } elseif (IPv4::isValid($ip)) {
-        $ip_address = \App\Models\Ipv4Address::query()
+        $ip_addresses = \App\Models\Ipv4Address::query()
             ->where('ipv4_address', $ip)
             ->with('port.device')
-            ->first();
+            ->get();
     }
+    if (isset($ip_addresses)) {
+        $ip_addresses = $ip_addresses->filter(function ($ip_address) {
+            return isset($ip_address->port);
+        });
+        if (! empty($within_poller_groups)) {
```

Comment:
You should do this in the query instead of in php.

Reply:
Was able to move the port existence check into the query, but when I did a double join on `ports` and `devices` along with the `with('port.device')` clause to check for thee poller group, I was able to get the correct IP address (when I print it I see both `port` and `device` in the JSON), but `$ip_address->port->device` gives me null.

Here's the old query with both port and poller group checks:
```
        $ip_addresses = \App\Models\Ipv4Address::query()
            ->where('ipv4_address', $ip)
            ->whereNotNull('ipv4_addresses.port_id')
            ->join('ports', 'ipv4_addresses.port_id', '=', 'ports.port_id')
            ->whereNotNull('ports.device_id')
            ->join('devices', 'ports.device_id', '=', 'devices.device_id')
            ->when(! empty($within_poller_groups), function($query) use ($within_poller_groups) {
                    return $query->whereIn('devices.poller_group', $within_poller_groups);
            })
            ->with('port.device')
            ->first();
```

Reply:
The joins break the eloquent loading.  You can check a relationship like this: (may not be 100% correct)
```
->whereHas('port.device.groups', function ($query) use ($within_poller_groups) {
    $query->whereIn('device_groups.id', $within_poller_groups);
})
```

Reply:
Got it, thanks. Since the join breaks eloquent loading can I keep the poller group check outside of the query?